python/ray/autoscaler/command_runner.py

```diff
@@ -326,6 +345,163 @@ def _set_ssh_ip_if_required(self):
+                                # is just starting up, so we want to handle
+                                # it on a case-by-case basis
+                                if failure_message is not None:
+                                    raise ValueError("Two errors is weird")
+
+                                failure_message="ssh_conn_refused"
+                                continue
+
+
+                            if self.known_host_update_msg_re.\
+                                fullmatch(l) is not None:
+                                # this is not an error at all, in fact we
+                                # completely ignore hosts files on purpose
+                                continue
+
+                            if dump_stderr:
+                                output_file.write(l+"\n")
+
+                            cli_logger.error(l)
+
+                return failure_message
+
+            with self.process_runner.Popen(
+                final_cmd,
+                # do not inherit stdin as it messes with signals
+                # (ctrl-C for SIGINT) and these commands aren't
+                # supposed to take input anyway
+                #
+                # P.S. it also messes up the terminal formatting
+                # probably due to some weird stuff that SSH
+                # does to stdin internally if it's a TTY
+                stdin=self.process_runner.PIPE,
+                # fixme: for some reason this is a tty which we DON'T want
+                # commands spam A LOT if they think it's a real tty
+                # especially for progress bars and such
+                # pip is a HUGE issue since it redraws the progress bar
+                # even if no progress was made
+                stdout=self.process_runner.PIPE,
+                stderr=self.process_runner.PIPE,
+                bufsize=1, # line buffering
+                universal_newlines=True # text mode outputs
+            ) as p:
+                from concurrent.futures import wait, ThreadPoolExecutor
+
+                # stdout + stderr
+                with ThreadPoolExecutor(max_workers=2) as pool:
+                    stdout_future = \
+                        pool.submit(read_stream,
+                            p.stdout, is_stdout=True)
+                    stderr_future = \
+                        pool.submit(read_stream,
+                            p.stderr, is_stdout=False)
+                    pool.shutdown() # wait for completion
+
+                    # do we kill? in theory the streams should have closed
+                    # and both threads finished
+                    # p.kill()
+
+                    p.poll()
+
+                    failure_message = stdout_future.result()
+                    if stderr_future.result() is not None:
+                        if failure_message is not None:
+                            raise ValueError("Two errors is weird")
```

Comment:
This is a poor error message. Could you provide some context for someone who sees this on the screen?

Reply:
Good catch, I'll change this.

This is an `assert False, "PANIC"` kind of scenario because unless there is a bug in the logic this should never-ever happen anyway.